pykafka/protocol.py

```diff
@@ -794,17 +814,45 @@ def _unpack_message_set(self, buff, partition_id=-1):
         for message in message_set.messages:
             if message.compression_type == CompressionType.NONE:
                 output.append(message)
+                continue
             elif message.compression_type == CompressionType.GZIP:
                 decompressed = compression.decode_gzip(message.value)
-                output += self._unpack_message_set(decompressed,
-                                                   partition_id=partition_id)
+                messages = self._unpack_message_set(decompressed,
+                                                    partition_id=partition_id)
             elif message.compression_type == CompressionType.SNAPPY:
                 decompressed = compression.decode_snappy(message.value)
-                output += self._unpack_message_set(decompressed,
-                                                   partition_id=partition_id)
+                messages = self._unpack_message_set(decompressed,
+                                                    partition_id=partition_id)
+            if messages[-1].offset < message.offset:
+                # With protocol 1, offsets from compressed messages start at 0
+                assert messages[0].offset == 0
+                delta = message.offset - len(messages) + 1
```

Comment:
Oh, that makes sense. My incorrect assumption was that the containing message has an offset equal to `previous_offset + 1` instead of what it apparently actually is, which is `previous_offset + len(internal_messages)`.

By the way, your help in figuring out what Kafka is actually doing here is invaluable.